File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/NoStartStateWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepMovement;

final class NoStartStateWorkflow implements Flow<Void> {
    static final long RPC_OUTPUT = 100L;
    private final TriggeredStep triggered = new TriggeredStep();

    @Override
    public StepList<Void> getSteps() {
        return StepList.withoutStartStep(triggered);
    }

    @RPC
    public RPCResult<Long> invoke(final Context context, final String input) {
        if (context.getFlowId().isEmpty() || context.getRunId().isEmpty()) {
            throw new IllegalStateException("invalid RPC context");
        }
        return RPCResult.of(RPC_OUTPUT, StepMovement.of(triggered, null));
    }

    static final class TriggeredStep implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            return StepDecision.gracefulComplete(1);
        }
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/no_start_state_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use dex_sdk::{
    Context, Flow, HandlerError, HandlerResult, Rpc, RpcList, RpcResult, Step, StepDecision,
    StepList, StepMovement,
};

pub(crate) struct NoStartStateWorkflow {
    triggered: TriggeredStep,
}

impl NoStartStateWorkflow {
    pub(crate) const RPC_OUTPUT: i64 = 100;
    pub(crate) const INVOKE: Rpc<String, i64> = Rpc::new("invoke");

    pub(crate) fn new() -> Self {
        Self {
            triggered: TriggeredStep,
        }
    }

    fn invoke(&self, context: &mut Context, _input: String) -> HandlerResult<RpcResult<i64>> {
        if context.flow_id().is_empty() || context.run_id().is_empty() {
            return Err(HandlerError::new(
                "NoStartStateFailure",
                "invalid RPC context",
            ));
        }
        Ok(RpcResult::new(Self::RPC_OUTPUT).then(StepMovement::to(&self.triggered, ())))
    }
}

impl Flow for NoStartStateWorkflow {
    type StartInput = ();

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::empty().and(&self.triggered)
    }

    fn rpcs(&self) -> RpcList<Self> {
        RpcList::new().function(Self::INVOKE, Self::invoke)
    }
}

struct TriggeredStep;

impl Step for TriggeredStep {
    type Input = ();

    fn execute(&self, _context: &mut Context, (): ()) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(1_i32))
    }
}
